metrics-api/src/examples/java/Sample.java

```diff
@@ -0,0 +1,64 @@
+import software.amazon.swage.collection.TypedMap;
+import software.amazon.swage.metrics.ContextData;
+import software.amazon.swage.metrics.Metric;
+import software.amazon.swage.metrics.MetricContext;
+import software.amazon.swage.metrics.ThreadLocalMetrics;
+import software.amazon.swage.metrics.Unit;
+import software.amazon.swage.metrics.record.MetricRecorder;
+import software.amazon.swage.metrics.record.NullRecorder;
+import software.amazon.swage.threadcontext.ThreadContext;
+
+/**
+ * Sample code that illustrates the different actors in the system.
+ */
+public class Sample {
+    private static final Metric NANO_TIME = Metric.define("Time");
+    private static final Metric CALL = Metric.define("Call");
+
+    public static void main(String[] args) {
+        // Role 1: Set up the metric recording system.
+        MetricRecorder recorder = new NullRecorder();
+
+        // Role 2: Set up the context in which measurements will be taken.
+        TypedMap metadata = ContextData.withId("id").build();
+        try (MetricContext context = recorder.context(metadata)) {
+
+            // Call a task that is aware of the MetricContext
+            contextAware(context);
+
+            // Options for calling code that is not aware of the metrics framework
+            // First we add the MetricContext to the ThreadContext that will be used when
+            // application code is called. Typically any other Thread-associated keys would
+            // be added at the same time.
+            ThreadContext threadContext = ThreadContext.emptyContext()
```

Comment:
There are separate roles here.

"Role 2" is framework code that is preparing to call application code and which is setting up multiple ThreadContext properties, metrics being only one of them. The code here is focus on setting up the thread local context rather than doing anything with metrics in particular. Once it has set up the ThreadContext it can reuse it for multiple calls into application code.

"Role 3" is library code that wants to emit metrics but which is called from application code that is unaware of the metrics framework being used. It relies on some higher level authority, in this example Role 2 above, to have provided ThreadLocalMetrics. Providing access to the "current" MetricContext allows it to pass that on to other Metrics-aware code, avoiding the need for a separate API like ScopedMetrics had.

ThreadLocalMetrics.current() means that any user (role 3) code emitting metrics won't need to change if we change how the context is bound to the thread. I'm not convinced we need to make the binding mechanism pluggable.